browser/brave_vpn/dns/brave_vpn_dns_observer_service.cc

```diff
@@ -0,0 +1,152 @@
+#include "brave/components/brave_vpn/brave_vpn_utils.h"
+#include "brave/components/brave_vpn/pref_names.h"
+#include "chrome/browser/net/secure_dns_config.h"
+#include "chrome/browser/net/stub_resolver_config_reader.h"
+#include "chrome/browser/net/system_network_context_manager.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "chrome/browser/ui/browser_window.h"
+#include "chrome/browser/ui/simple_message_box.h"
+#include "chrome/common/pref_names.h"
+#include "chrome/grit/chromium_strings.h"
+#include "components/grit/brave_components_strings.h"
+#include "components/prefs/pref_service.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace brave_vpn {
+
+namespace {
+const char kCloudflareDnsProviderURL[] =
+    "https://chrome.cloudflare-dns.com/dns-query";
+
+void SkipDNSDialog(PrefService* prefs, bool checked) {
+  if (!prefs)
+    return;
+  prefs->SetBoolean(prefs::kBraveVPNShowDNSPolicyWarningDialog, !checked);
+}
+
+gfx::NativeWindow GetAnchorBrowserWindow() {
+  auto* browser = chrome::FindLastActive();
+  return browser ? browser->window()->GetNativeWindow()
+                 : gfx::kNullNativeWindow;
+}
+
+bool AreConfigsEqual(SecureDnsConfig& one, SecureDnsConfig& two) {
+  return one.mode() == two.mode() &&
+         one.management_mode() == two.management_mode() &&
+         one.doh_servers() == two.doh_servers();
+}
+
+}  // namespace
+
+BraveVpnDnsObserverService::BraveVpnDnsObserverService(
+    PrefService* local_state,
+    PrefService* profile_prefs)
+    : local_state_(local_state), profile_prefs_(profile_prefs) {
+  DCHECK(profile_prefs_);
+  DCHECK(local_state_);
+  pref_registrar_.Init(local_state_);
+}
+
+BraveVpnDnsObserverService::~BraveVpnDnsObserverService() = default;
+
+void BraveVpnDnsObserverService::ShowPolicyWarningMessage() {
+  if (!profile_prefs_->GetBoolean(prefs::kBraveVPNShowDNSPolicyWarningDialog)) {
+    return;
+  }
+
+  if (policy_callback_) {
+    std::move(policy_callback_).Run();
+    return;
+  }
+
+  chrome::ShowWarningMessageBoxWithCheckbox(
+      GetAnchorBrowserWindow(), l10n_util::GetStringUTF16(IDS_PRODUCT_NAME),
+      l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_POLICY_ALERT),
+      l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_POLICY_CHECKBOX),
+      base::BindOnce(&SkipDNSDialog, profile_prefs_));
+}
+
+void BraveVpnDnsObserverService::ShowVpnNotificationDialog() {
+  if (dialog_callback_) {
+    dialog_callback_.Run();
+    return;
+  }
+  VpnNotificationDialogView::Show(chrome::FindLastActive());
+}
+
+void BraveVpnDnsObserverService::UnlockDNS() {
+  pref_registrar_.RemoveAll();
+  local_state_->ClearPref(::prefs::kBraveVpnDnsConfig);
+  // Read DNS config to initiate update of actual state.
+  SystemNetworkContextManager::GetStubResolverConfigReader()
+      ->UpdateNetworkService(false);
+}
+
+// Users are able to change the DNS mode when VPN is enabled if the dns settings
```

Comment:
if the user changes their DNS setting while VPN is enabled, it will automatically trigger the override because we always write to `kBraveVpnDnsConfig` even when it isn't used

Reply:
decided to remove it and lock settings each time, done in https://github.com/brave/brave-core/pull/15989/commits/9eb6cd0e0a671120745dc551ff14bf366bc4611a